src/plugins/discover/public/application/services_provider.tsx

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React from 'react';
+import { KibanaContextProvider } from '@kbn/kibana-react-plugin/public';
+import { CellActionsProvider } from '@kbn/cell-actions';
+import { DiscoverServices } from '../build_services';
+
+export const ServicesContextProvider: React.FC<{ services: DiscoverServices }> = ({
+  services,
+  children,
+}) => {
+  const { getTriggerCompatibleActions } = services.uiActions;
+  return (
+    <KibanaContextProvider services={services}>
+      <CellActionsProvider getTriggerCompatibleActions={getTriggerCompatibleActions}>
```

Comment:
@jughosta Yes that makes sense, I moved the `CellActionsProvider` inside the `DiscoverDocuments` component wrapping only the `DiscoverGrid`. This new services_provider is no longer needed, removed.

Reply:
@jughosta I had to use the provider inside `src/plugins/discover/public/application/context/context_app_content.tsx` as well

Reply:
There is also an option to defined a default value for the context. Example:

```typescript
const DEFAULT_ACTIONS = []
const CellActionsContext = createContext<{ getActions: GetActions }>({
  // set a default handler for cases when provider is not used
  getActions: () => Promise.resolve(DEFAULT_ACTIONS),
});

export const CellActionsProvider: FC<CellActionsProviderProps> = ({
  children,
  getTriggerCompatibleActions,
}) => {
  const value: { getActions: GetActions } = useMemo(
    () => ({
      getActions: (context) =>
        getTriggerCompatibleActions(context.trigger.id, context).then((actions) =>
          orderBy(['order', 'id'], ['asc', 'asc'], actions)
        ) as Promise<CellAction[]>,
    }),

    [getTriggerCompatibleActions]
  );

  // make sure that provider's value does not change
  return <CellActionsContext.Provider value={value}>{children}</CellActionsContext.Provider>;
};

// don't trigger an exception
export const useCellActionsContext = () => useContext(CellActionsContext);
```

Reply:
That's smart. 
But doing this would hide incorrect uses of the `CellActions` framework, making it harder to debug as well. The exception is intended, to make sure `CellActions` are only used inside its Provider, so we notice if missing during the development stage. We already had this discussion within the team and we decided to be strict.
(That's also why I put it at the top level from the beginning)